design-proposals/network-device-capacity.md

```diff
@@ -0,0 +1,98 @@
+# Network Devices Capacity
+
+Author: [Edward Haas](edwardh@redhat.com)
+Contributors: TBD
+State: Draft
+
+## Summary
+Support the ability to define how many network devices a VM may potentially contain.
+Allowing the use of hotplug to add network interfaces up to that capacity limit.
+
+## Motivation
+The introduction of network devices hotplug introduced a limitation
+to the amount of possible devices one can attach.
+
+Users of libvirt who can estimate how many potential network devices are needed,
+can define the additional controllers in advance.
+KubeVirt does not expose such domain details through its API.
+But even if these details had been exposed by the VMI object,
+a common user would find it challenging to define such low domain details.
+
+### Goals
+- Allow users to define how many network devices can be potentially hotplug.
+- Do not expose domain architecture details (e.g. PCI/PCIe controllers).
+
+## Proposal
+Users are to express the potential amount of network devices through
+the VMI spec.
+Internally, kubevirt will use libvirt logic to generate the needed
+devices to support the amount specified (e.g. PCIe controllers).
+
+Using hotplug, users may add network devices up to the amount specified
+by the new user *interface-capacity* input.
+
+### Definition of Personas
+- VM/VMI administrator: Any user who can create VM and VMI objects.
+
+### User Stories
+- TBD
+
+### API Extensions
+The VMI Spec will be extended to include a new field, named `interfaceCapacity`.
```

Comment:
IMO, this name is bad. I assume I'm specifying the capacity of an interface, not the max number of interfaces available for a VM.

Reply:
I think the location in the spec does give it a context, so it is not that bad.
But if you have something to suggest, feel free to propose it.

I though also about `interfacesCountCapacity` but I liked it less.

Reply:
This is an API (which pretty much ends up living forever), let's make the extra effort of choosing something that is crystal clear. 

I even agree with your point: the location within the spec does give it some context. But still ... in yaml it is *very* easy to get these things wrong. 

I'll give it some thought, and hopefully I can suggest something more palatable.

Reply:
I think `interfacesListCapacity` would be a better name, as it more clearly indicates that it's related to the `interface` list and would naturally sort close to it.

I know your focus is mainly on network interfaces, but really the logic is going to be the same for other types of PCI-connected devices such as disks. So by adopting this naming convention we would have

```yaml
domain:
  devices:
    disks:
      - disk:
          bus: virtio
        name: containerdisk
    disksListCapacity: 2
    interfaces:
      - masquerade: {}
        name: default
      - bridge: {}
        name: secondary
    interfacesListCapacity: 5
```

Alternative structure:

```yaml
domain:
  devices:
    disks:
      - disk:
          bus: virtio
        name: containerdisk
    interfaces:
      - masquerade: {}
        name: default
      - bridge: {}
        name: secondary
  devicesListCapacity:
    disks: 2
    interfaces: 5
```

This has the advantage of keeping "meta-information about lists of devices" clearly separated from "actual lists of devices", and has a nice mirroring property to it. On the other hand, keeping all information related to each list (number of allowed elements, current contents) close together might be preferable.